### PR TITLE
[RUST] python (re)integration v1

### PR DESCRIPTION
**Pull request by changhiskhan**
1. exposes `lance.write_dataset` and `lance.dataset`.
2. Has APIs for projection pushdown, offset/limit, and nearest neighbors
3. No predicate pushdown yet
4. The dataset is not readable by duckdb directly. Requires calling either `to_table()` or `to_scanner().to_reader()` (like datafusion or polars)

### CODE REVIEW

File: rust/src/dataset/scanner.rs [hunk 1]
```diff
@@ -24,6 +24,8 @@ use arrow_schema::{Schema as ArrowSchema, SchemaRef};
 use futures::stream::Stream;
 use futures::StreamExt;
 use object_store::path::Path;
+use tokio::sync::mpsc::{self, Receiver};
```
Comment (eddyxu) [inline rust/src/dataset/scanner.rs:27]: are these useful?

File: rust/pylance/python/lance/dataset.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    max_rows_per_group: int, default 1024
+        The max number of rows before starting a new group (in the same file)
+    """
+    if isinstance(data_obj, pa.Table):
+        reader = data_obj.to_reader()
+    elif isinstance(data_obj, pa.dataset.Dataset):
+        reader = pa.dataset.Scanner.from_dataset(data_obj).to_reader()
+    elif isinstance(data_obj, pa.dataset.Scanner):
+        reader = data_obj.to_reader()
+    elif isinstance(data_obj, pa.RecordBatchReader):
```
Comment (eddyxu) [inline rust/pylance/python/lance/dataset.py:124]: This is nice. Finally have the generator support. Could we add a test for it?

File: rust/pylance/src/scanner.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+#[pymethods]
+impl Scanner {
+    #[getter(dataset_schema)]
+    fn dataset_schema(self_: PyRef<'_, Self>) -> PyResult<PyObject> {
+        ArrowSchema::from(self_.dataset.schema()).to_pyarrow(self_.py())
+    }
+
+    fn to_reader(self_: PyRef<'_, Self>) -> PyResult<PyObject> {
+        self_.rt.block_on(async {
+            let mut scanner: LanceScanner = self_.dataset.scan();
```
Comment (eddyxu) [inline rust/pylance/src/scanner.rs:72]: did not see which of the following code needs to be async. 

is it ok to not use `rt.block_on`?

File: rust/pylance/src/scanner.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            limit,
+            rt,
+        }
+    }
+}
+
+#[pymethods]
+impl Scanner {
+    #[getter(dataset_schema)]
+    fn dataset_schema(self_: PyRef<'_, Self>) -> PyResult<PyObject> {
```
Comment (eddyxu) [inline rust/pylance/src/scanner.rs:66]: why use `self_`  as name?

File: rust/pylance/src/scanner.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        ArrowSchema::from(self_.dataset.schema()).to_pyarrow(self_.py())
+    }
+
+    fn to_reader(self_: PyRef<'_, Self>) -> PyResult<PyObject> {
```
Comment (eddyxu) [inline rust/pylance/src/scanner.rs:70]: why use `self_` as name.

File: rust/pylance/src/scanner.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use pyo3::prelude::*;
+use pyo3::{pyclass, PyObject, PyResult};
+/// This a Scanner implementation which will being associated
+/// with a Pyarrow Scanner so duckdb can be tricked to call
+/// it's methods during scans
+use std::sync::Arc;
```
Comment (eddyxu) [inline rust/pylance/src/scanner.rs:23]: Lets order uses in section? `std` / `thirdparty` / `lance`?

File: rust/pylance/src/reader.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+impl Iterator for LanceReader {
+    type Item = Result<RecordBatch, ArrowError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let stream = &mut self.stream;
+        self.rt.block_on(async {
+            stream.next().await.map(|rs| {
+                rs.map_err(|err| {
+                    match err {
+                        Error::Arrow(err) => ArrowError::IoError(err), // we lose the error type converting to LanceError
```
Comment (eddyxu) [inline rust/pylance/src/reader.rs:55]: Should we implement a `From<&LanceError> for ArrowError`

File: rust/pylance/src/dataset.rs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        arrow_schema.to_pyarrow(self_.py())
+    }
+
+    fn scanner(
+        &self,
+        columns: Option<Vec<String>>,
+        limit: i64,
+        offset: Option<i64>,
+    ) -> PyResult<Scanner> {
+        println!("{:?}", columns);
```
Comment (eddyxu) [inline rust/pylance/src/dataset.rs:71]: Let's remove prints.

File: rust/pylance/src/scanner.rs (hunk 3 above)
Comment (changhiskhan) [reply rust/pylance/src/scanner.rs:72]: into_stream calls tokio spawn so requires a Runtime.

File: rust/pylance/src/scanner.rs (hunk 4 above)
Comment (changhiskhan) [reply rust/pylance/src/scanner.rs:66]: Convention from examples - this has a reference to Python whereas &self: Self does not

File: rust/pylance/src/scanner.rs (hunk 6 above)
Comment (changhiskhan) [reply rust/pylance/src/scanner.rs:23]: is there a cargo fmt setting for this?

File: rust/src/dataset/scanner.rs (hunk 1 above)
Comment (changhiskhan) [reply rust/src/dataset/scanner.rs:27]: I think this was rebase error.

File: rust/pylance/src/scanner.rs (hunk 3 above)
Comment (changhiskhan) [reply rust/pylance/src/scanner.rs:72]: https://github.com/eto-ai/lance/blob/main/rust/src/io/exec/scan.rs#L55

File: rust/src/dataset/scanner.rs [hunk 9]
```diff
@@ -88,21 +90,37 @@ impl<'a> Scanner {
...
         self.limit = Some(limit);
         self.offset = offset;
-        self
+        Ok(self)
     }
 
     /// Find k-nearest neighbour within the vector column.
-    pub fn nearest(&mut self, column: &str, q: &Float32Array, k: usize) -> &mut Self {
+    pub fn nearest(&mut self, column: &str, q: &Float32Array, k: usize) -> Result<&mut Self> {
+        if k <= 0 {
```
Comment (eddyxu) [inline rust/src/dataset/scanner.rs:109]: k is `usize` which can not be negative, it only needs to check `k == 0`?

**Review by eddyxu — APPROVED** [review]
Go go go

File: rust/pylance/python/lance/__init__.py [hunk 10]
```diff
@@ -0,0 +1,6 @@
+from .dataset import LanceDataset, write_dataset
```
Comment (eddyxu) [inline rust/pylance/python/lance/__init__.py:1]: Not urgent not. but let's add license headers later?

**Pull request merged**